bridge_networkstatus.go

```diff
@@ -0,0 +1,214 @@
+}
+
+type BridgeNetworkStatus struct {
+	RouterStatuses map[Fingerprint]GetStatus
+}
+
+// Iterate implements the ObjectSet interface.  Using a channel, it iterates
+// over and returns all router statuses.  The given object filter can be used
+// to filter router statuses, e.g., by fingerprint.
+func (bn *BridgeNetworkStatus) Iterate(filter *ObjectFilter) <-chan Object {
+
+	ch := make(chan Object)
+
+	go func() {
+		for _, getStatus := range bn.RouterStatuses {
+			status := getStatus()
+			if filter == nil || filter.IsEmpty() || filter.MatchesRouterStatus(status) {
+				ch <- status
+			}
+		}
+		close(ch)
+	}()
+
+	return ch
+}
+
+// parseBridgeStatusUnchecked parses a descriptor of type
+// "bridge-network-status".  The input should be without a type annotation;
+// i.e., the type annotation should already have been read and checked to be the
+// correct type.  The function returns a network consensus if parsing was
+// successful.  If there were any errors, an error string is returned.  If the
+// lazy argument is set to true, parsing of the router statuses is delayed until
+// they are accessed.
+func parseBridgeStatusUnchecked(r io.Reader, lazy bool) (*BridgeNetworkStatus, error) {
+
+	var networkstatus = BridgeNetworkStatus{RouterStatuses: make(map[Fingerprint]GetStatus)}
+	var statusParser func(string) (Fingerprint, GetStatus, error)
+
+	if lazy {
+		statusParser = LazyParseRawStatus
+	} else {
+		statusParser = ParseRawStatus
+	}
+
+	// We will read raw router statuses from this channel.
+	queue := make(chan QueueUnit)
+	go DissectFile(r, extractBridgeStatusEntry, queue)
+
+	// Parse incoming router statuses until the channel is closed by the remote
+	// end.
+	for unit := range queue {
+		if unit.Err != nil {
+			return nil, unit.Err
+		}
+
+		fingerprint, getStatus, err := statusParser(unit.Blurb)
+		if err != nil {
+			return nil, err
+		}
+
+		networkstatus.RouterStatuses[SanitiseFingerprint(fingerprint)] = getStatus
+	}
+
+	return &networkstatus, nil
+}
+
+// extractBridgeStatusEntry is a bufio.SplitFunc that extracts individual network
+// status entries.
+func extractBridgeStatusEntry(data []byte, atEOF bool) (advance int, token []byte, err error) {
+
+	if atEOF && len(data) == 0 {
+		return 0, nil, nil
+	}
+
+	start := 0
+	if !bytes.HasPrefix(data, []byte("r ")) {
+		start = bytes.Index(data, []byte("\nr "))
+		if start < 0 {
+			if atEOF {
+				return 0, nil, fmt.Errorf("cannot find beginning of status entry: \"\\nr \"")
+			}
+			// Request more data.
+			return 0, nil, nil
+		}
+		start++
+	}
+
+	end := bytes.Index(data[start:], []byte("\nr "))
+	if end >= 0 {
+		return start + end + 1, data[start : start+end+1], nil
+	}
+	end = bytes.Index(data[start:], []byte("directory-signature"))
+	if end >= 0 {
+		// "directory-signature" means this is the last status; stop
+		// scanning.
+		return start + end, data[start : start+end], bufio.ErrFinalToken
+	}
+	if atEOF {
+		return len(data), data[start:], nil
```

Comment:
This function is exactly the same than `extractStatusEntry` except of this line. I thought about changing `extractStatusEntry`, but I was not sure if the error was important for the consensus file.